### ISSUE TITLE
LLVM assertion on cross compile to x86_64-pc-windows-msvc

### ISSUE THREAD

**Issue by Kimundi** [rust-lang/rust#33463]
Attempting to cross compile from 64Bit Arch after doing

```
rustup target add x86_64-pc-windows-msvc
cargo build --release --target=x86_64-pc-windows-msvc --verbose
```

Errors with LLVM asserts:

```
marvin@tp-marvin-arch ~/d/r/g/fun1> cargo build --release --target=x86_64-pc-windows-msvc --verbose
       Fresh tuple_utils v0.2.0
       Fresh winapi v0.2.6
       Fresh fnv v1.0.2
       Fresh atom v0.3.4
   Compiling threadpool v1.0.0
     Running `rustc /home/marvin/.cargo/registry/src/github.com-88ac128001ac3a9a/threadpool-1.0.0/src/lib.rs --crate-name threadpool --crate-type lib -C opt-level=3 -C metadata=0bb5a6ba1d0b4f62 -C extra-filename=-0bb5a6ba1d0b4f62 --out-dir /home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --emit=dep-info,link --target x86_64-pc-windows-msvc -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --cap-lints allow`
       Fresh gdi32-sys v0.1.1
       Fresh lazy_static v0.1.16
       Fresh libc v0.2.11
   Compiling rustc-serialize v0.3.19
     Running `rustc /home/marvin/.cargo/registry/src/github.com-88ac128001ac3a9a/rustc-serialize-0.3.19/src/lib.rs --crate-name rustc_serialize --crate-type lib -C opt-level=3 -C metadata=3bc953984ed46e7f -C extra-filename=-3bc953984ed46e7f --out-dir /home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --emit=dep-info,link --target x86_64-pc-windows-msvc -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --cap-lints allow`
   Compiling shared_library v0.1.4
     Running `rustc /home/marvin/.cargo/registry/src/github.com-88ac128001ac3a9a/shared_library-0.1.4/src/lib.rs --crate-name shared_library --crate-type lib -C opt-level=3 -C metadata=8a12fcda5e4d258b -C extra-filename=-8a12fcda5e4d258b --out-dir /home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --emit=dep-info,link --target x86_64-pc-windows-msvc -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --extern lazy_static=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps/liblazy_static-5e6d9f365bf63baa.rlib --extern libc=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps/liblibc-38919d24e617a235.rlib --cap-lints allow`
   Compiling rand v0.3.14
     Running `rustc /home/marvin/.cargo/registry/src/github.com-88ac128001ac3a9a/rand-0.3.14/src/lib.rs --crate-name rand --crate-type lib -C opt-level=3 -C metadata=c724acb3942597d1 -C extra-filename=-c724acb3942597d1 --out-dir /home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --emit=dep-info,link --target x86_64-pc-windows-msvc -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --extern libc=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps/liblibc-38919d24e617a235.rlib --cap-lints allow`
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/include/llvm/ADT/DenseMap.h:1025: llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type* llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::operator->() const [with KeyT = const llvm::MachineBasicBlock*; ValueT = int; KeyInfoT = llvm::DenseMapInfo<const llvm::MachineBasicBlock*>; Bucket = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>; bool IsConst = false; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::pointer = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>*; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>]: Assertion `isHandleInSync() && "invalid iterator access!"' failed.
Build failed, waiting for other jobs to finish...
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/include/llvm/ADT/DenseMap.h:1025: llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type* llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::operator->() const [with KeyT = const llvm::MachineBasicBlock*; ValueT = int; KeyInfoT = llvm::DenseMapInfo<const llvm::MachineBasicBlock*>; Bucket = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>; bool IsConst = false; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::pointer = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>*; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>]: Assertion `isHandleInSync() && "invalid iterator access!"' failed.
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/include/llvm/ADT/DenseMap.h:1025: llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type* llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::operator->() const [with KeyT = const llvm::MachineBasicBlock*; ValueT = int; KeyInfoT = llvm::DenseMapInfo<const llvm::MachineBasicBlock*>; Bucket = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>; bool IsConst = false; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::pointer = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>*; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>]: Assertion `isHandleInSync() && "invalid iterator access!"' failed.
rustc: /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/include/llvm/ADT/DenseMap.h:1025: llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type* llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::operator->() const [with KeyT = const llvm::MachineBasicBlock*; ValueT = int; KeyInfoT = llvm::DenseMapInfo<const llvm::MachineBasicBlock*>; Bucket = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>; bool IsConst = false; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::pointer = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>*; llvm::DenseMapIterator<KeyT, ValueT, KeyInfoT, Bucket, IsConst>::value_type = llvm::detail::DenseMapPair<const llvm::MachineBasicBlock*, int>]: Assertion `isHandleInSync() && "invalid iterator access!"' failed.
error: Could not compile `shared_library`.

Caused by:
  Process didn't exit successfully: `rustc /home/marvin/.cargo/registry/src/github.com-88ac128001ac3a9a/shared_library-0.1.4/src/lib.rs --crate-name shared_library --crate-type lib -C opt-level=3 -C metadata=8a12fcda5e4d258b -C extra-filename=-8a12fcda5e4d258b --out-dir /home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --emit=dep-info,link --target x86_64-pc-windows-msvc -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps -L dependency=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps --extern lazy_static=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps/liblazy_static-5e6d9f365bf63baa.rlib --extern libc=/home/marvin/dev/rust/graphics_fun_2016-05-05/fun1/target/x86_64-pc-windows-msvc/release/deps/liblibc-38919d24e617a235.rlib --cap-lints allow` (exit code: 1)

```

Compiler version:

```
rustc 1.10.0-nightly (8da2bcac5 2016-04-28)
binary: rustc
commit-hash: 8da2bcac5db1e091b90cceb19d0496f0f7501c88
commit-date: 2016-04-28
host: x86_64-unknown-linux-gnu
release: 1.10.0-nightly
```

**Comment by alexcrichton** [comment]
I've personally run into this quite a few times locally as well, but I was testing awhile back and found that llvm-mirror/llvm@285cad0e0ce95c97d8834e09f4b98942bfebdc32 fixed it. We'll pick it up on the next LLVM upgrade

**Comment by nagisa** [comment]
Also hit this. Need to upgrade LLVM faster.

**Comment by DanielKeep** [comment]
Hit this on 1.12 nightly 9316ae515, hosted on i686-pc-windows-gnu.

**Comment by lilianmoraru** [comment]
https://github.com/rust-lang/rust/blob/1572bf104dbf65d58bd6b889fa46229c9b92d6f9/src/libstd/sys/mod.rs#L37 - it's not using `target_os`, shouldn't this block cross-compilation?

**Comment by Mark-Simulacrum** [comment]
I think the `target_os` issue (if it's an issue, which I hope it isn't...) isn't part of this bug. Otherwise, this should've been fixed by the LLVM 4.0 upgrade, so I'm going to presume fixed and close. The commit noted above is included in the LLVM currently in-tree.
